### PR TITLE
Handle the case when application has no focus in ApplicationPartService

### PR DESCRIPTION
**Pull request by laeubi**
Currently when the application has no focus there is also no active window child and in this case several actions still fail even though a part is available for perform actions.

This now handles the case of an application without focus with an empty optional to allow fall back to the part context. If that context has still no suitable part service implementation as a last resort the containing window context is used.

### CODE REVIEW

File: tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java [hunk 1]
```diff
@@ -134,11 +137,24 @@ private void removeAllWorkingSets() {
...
+			}
+			try {
+				root.refreshLocal(IResource.DEPTH_INFINITE, null);
+				if (root.exists()) {
+					root.delete(true, null);
+				}
+			} catch (CoreException e1) {
+				TestPlugin.getDefault().getLog().log(e.getStatus());
+				throw createAssertionError(e);
+			}
```
Comment (fedejeanne) [inline tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java]: Are these changes related to this PR? Judging by the description of the PR, it doesn't look like they do.

**Review by fedejeanne — DISMISSED** [review]
@laeubi thank you for contributing this. I only have a minor concern regarding the changes in the test class.

NOTE: I only skimmed through the the other changes but I haven't tested them. I could test them do it if you briefly describe how to do it.

File: tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java (hunk 1 above)
Comment (laeubi) [reply tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java]: The test seem to randomly fail so I added another commit here to see if it can be fixed.

Comment (fedejeanne) [reply tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java]: But the failing test doesn't have anything to do with `ApplicationPartServiceImpl` right?

Comment (laeubi) [reply tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java]: Only that it failed while building the PR once...

Comment (HeikoKlare) [reply tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java]: I would propose to keep this separate and have another PR that is also linked to the according issue: #275

Comment (fedejeanne) [reply tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/dialogs/UIWorkingSetWizardsAuto.java]: Ah, thank you @HeikoKlare , I was looking for the right issue to link here :-)

@laeubi please propose another PR for these changes.

**Review by fedejeanne — COMMENTED** [review]
Looks better. I'm still missing a brief "How to test" in this PR and a description of the original bug/behavior that caused this PR.

**Comment by laeubi** [comment]
Sadly the case is quite tricky to test, you need a custom RCP application, you need an external trigger (e.g. Network communication) ...

**Comment by laeubi** [comment]
One additional test failure reported here:

- https://github.com/eclipse-platform/eclipse.platform.ui/issues/2649

but it suceeds now.

The other test failure is already known:

- https://github.com/eclipse-platform/eclipse.platform.ui/issues/294
